eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php

```diff
@@ -61,13 +84,230 @@ protected function configure()
+            $this->searchIndexer->updateSearchIndex($contentIds, $commit);
+            $progress->advance(1);
+        }
+
+        $progress->finish();
+    }
+
+    private function runParallelProcess(ProgressBar $progress, Statement $stmt, $processCount, $iterationCount, $commit)
+    {
+        /**
+         * @var \Symfony\Component\Process\Process[]|null[]
+         */
+        $processes = array_fill(0, $processCount, null);
+        $generator = $this->fetchIteration($stmt, $iterationCount);
+        do {
+            foreach ($processes as $key => $process) {
+                if ($process !== null && $process->isRunning()) {
+                    continue;
+                }
+
+                if ($process !== null) {
+                    // One of the processes just finished, so we increment progress bar
+                    $progress->advance(1);
+                }
+
+                if (!$generator->valid()) {
+                    unset($processes[$key]);
+                    continue;
+                }
+
+                $processes[$key] = $this->getPhpProcess($generator->current(), $commit);
+                $processes[$key]->start();
+                $generator->next();
+            }
+
+            if (!empty($processes)) {
+                sleep(1);
+            }
+        } while (!empty($processes));
+
+        $progress->finish();
+    }
+
+    /**
+     * @param DateTime $since
+     * @param string $select Field to select, one of  'c.id' and 'count(c.id)'
+     *
+     * @return \Doctrine\DBAL\Driver\Statement
+     */
+    private function getStatementContentSince(DateTime $since, $select = 'c.id')
+    {
+        $q = $this->connection->createQueryBuilder()
+            ->select($select)
+            ->from('ezcontentobject', 'c')
+            ->where('c.status = :status')->andWhere('c.modified >= :since')
+            ->orderBy('c.modified')
+            ->setParameter('status', ContentInfo::STATUS_PUBLISHED, PDO::PARAM_INT)
+            ->setParameter('since', $since->getTimestamp(), PDO::PARAM_INT);
+
+        return $q->execute();
+    }
+
+    /**
+     * @param string $select Field to select, one of  'c.id' and 'count(c.id)'
+     *
+     * @return \Doctrine\DBAL\Driver\Statement
+     */
+    private function getStatementContentAll($select = 'c.id')
+    {
+        $q = $this->connection->createQueryBuilder()
+            ->select($select)
+            ->from('ezcontentobject', 'c')
+            ->where('c.status = :status')
+            ->setParameter('status', ContentInfo::STATUS_PUBLISHED, PDO::PARAM_INT);
+
+        return $q->execute();
+    }
+
+    /**
+     * @param \Doctrine\DBAL\Driver\Statement $stmt
+     * @param int $iterationCount
+     *
+     * @return \Generator Return an array of arrays, each array contains content id's of $iterationCount.
+     */
+    private function fetchIteration(Statement $stmt, $iterationCount)
```

Comment:
Now I see why you need these Statements here :)
But maybe this could also be moved to Gateway, but using more domain-specific naming and internally reusing Statement methods inside GW?